File: 54-Spiral Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix or not matrix[0]:
            return []
        
        row = len(matrix)
        col = len(matrix[0])

        left, right, up, down = 0, col-1, 0, row-1
        res = []
        while True:
            if left<=right:
                for i in range(left, right+1):
                    res.append(matrix[up][i])
            else:
                break
            
            up+=1
            
            if up <= down:
                for i in range(up,down+1):
                    res.append(matrix[i][right])
            else:
                break
                
            right-=1
            
            if left <= right:
                for i in range(right, left-1, -1):
                    res.append(matrix[down][i])
            else:
                break
                
            down-=1
            
            if up<=down:
                for i in range(down, up-1,-1):
                    res.append(matrix[i][left])
            else:
                break
                
            left += 1
                
        return res

File: JavaScript/54-Spiral Matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    if (!matrix.length || !matrix[0].length ) return [];
    let left = 0, right = matrix[0].length-1, up = 0, down = matrix.length-1;
    let res = [];
    while (true){
        for (l = left; l <= right; l++){
            res.push(matrix[up][l]);
        }
        if (++up > down) break;
        
        for (u = up; u <= down; u++){
            res.push(matrix[u][right]);
        }
        if (--right < left) break;
        
        for (r = right; r >= left; r--){
            res.push(matrix[down][r]);
        }
        if (--down < up) break;
        
        for (d = down; d >= up; d--){
            res.push(matrix[d][left]);
        }
        if (++left > right) break;
    }
    return res;
};